addon/bootstrap.js

```diff
@@ -30,15 +39,31 @@ const prefObserver = {
     // aSubject is the nsIPrefBranch we're observing (after appropriate QI)
     // aData is the name of the pref that's been changed (relative to aSubject)
     if (aData == USER_DISABLE_PREF || aData == SYSTEM_DISABLE_PREF) {
-      handleStartup();
+      appStartupPromise.then(handleStartup);
     }
   }
 };
 
+const appStartupObserver = {
+  register: function() {
+    Services.obs.addObserver(this, "sessionstore-windows-restored", false);
+  },
+
+  unregister: function() {
+    Services.obs.removeObserver(this, "sessionstore-windows-restored", false);
+  },
+
+  observe: function() {
+    appStartupDone();
+    this.unregister();
+  }
+}
+
 function startup(data, reason) { // eslint-disable-line no-unused-vars
+  appStartupObserver.register();
```

Comment:
Are we confident that there's no way this code could be called after sessionstore-windows-restored is fired?  Like... if the add-on is reloaded?